src/main/java/com/aws/greengrass/clientdevices/auth/iot/GreengrassClientFactory.java

```diff
@@ -0,0 +1,135 @@
+ */
+
+package com.aws.greengrass.clientdevices.auth.iot;
+
+import com.aws.greengrass.componentmanager.ClientConfigurationUtils;
+import com.aws.greengrass.deployment.DeviceConfiguration;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.tes.LazyCredentialProvider;
+import com.aws.greengrass.util.Coerce;
+import com.aws.greengrass.util.GreengrassServiceClientFactory;
+import com.aws.greengrass.util.IotSdkClientFactory;
+import com.aws.greengrass.util.RegionUtils;
+import com.aws.greengrass.util.Utils;
+import com.aws.greengrass.util.exceptions.InvalidEnvironmentStageException;
+import software.amazon.awssdk.core.client.config.ClientOverrideConfiguration;
+import software.amazon.awssdk.core.retry.RetryMode;
+import software.amazon.awssdk.core.retry.RetryPolicy;
+import software.amazon.awssdk.http.apache.ApacheHttpClient;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.greengrassv2.GreengrassV2Client;
+import software.amazon.awssdk.services.greengrassv2.GreengrassV2ClientBuilder;
+import software.amazon.awssdk.services.greengrassv2data.GreengrassV2DataClient;
+import software.amazon.awssdk.services.greengrassv2data.GreengrassV2DataClientBuilder;
+
+import java.net.URI;
+import java.util.Optional;
+import javax.inject.Inject;
+
+/**
+ * Provides GG service clients built specifically for CDA use case.
+ * Ideally this factory should maintain the lifecycle of its clients,
+ * but for now its upto the consumer of the clients to maintain their lifecycle.
+ */
+public class GreengrassClientFactory {
+    private static final Logger logger = LogManager.getLogger(GreengrassClientFactory.class);
+    private final GreengrassServiceClientFactory greengrassServiceClientFactory;
+    private final LazyCredentialProvider lazyCredentialProvider;
+
+    /**
+     * Construct Greengrass Client Factory.
+     *
+     * @param greengrassServiceClientFactory GG Service Client Factory to retrieve the latest device-config
+     * @param lazyCredentialProvider Credential provider for the client
+     */
+    @Inject
+    public GreengrassClientFactory(GreengrassServiceClientFactory greengrassServiceClientFactory,
+                                   LazyCredentialProvider lazyCredentialProvider) {
+        this.greengrassServiceClientFactory = greengrassServiceClientFactory;
+        this.lazyCredentialProvider = lazyCredentialProvider;
+    }
+
+    /**
+     * Provides a new GG v2 client.
+     * Its upto the caller to maintain the lifecycle of the client.
+     * (TODO: This should ideally exist in Nucleus's GreengrassServiceClientFactory)
+     *
+     * @return GreengrassV2Client
+     */
+    public GreengrassV2Client getGGV2Client()  {
+        DeviceConfiguration deviceConfiguration = getDeviceConfiguration();
+        ApacheHttpClient.Builder httpClient = ClientConfigurationUtils
+                .getConfiguredClientBuilder(deviceConfiguration);
```

Comment:
this method gives you a client which is configured for tls mutual auth. That is not what you want for using the control plane.

Should use `getSdkHttpClientBuilder` or `getSdkHttpClient`

Reply:
We should essentially just be copy/pasting what Nucleus is doing here, right? Except skip the retry policy

Reply:
this is for the control plane client, skipped it from this PR.